From 4717657053a13d69d48dc82dbc5402f6712f66ea Mon Sep 17 00:00:00 2001
From: Erwan Tulou <brezhoneg1@yahoo.fr>
Date: Fri, 5 Jun 2009 09:55:09 +0200
Subject: [PATCH] QT4 menus : remove previous signal-slot connection(s) if any (correct trac #2818)

---
 modules/gui/qt4/menus.cpp |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/modules/gui/qt4/menus.cpp b/modules/gui/qt4/menus.cpp
index 0a12c20..9304752 100644
--- a/modules/gui/qt4/menus.cpp
+++ b/modules/gui/qt4/menus.cpp
@@ -1386,6 +1386,10 @@ void QVLCMenu::CreateAndConnect( QMenu *menu, const char *psz_var,
 
     MenuItemData *itemData = new MenuItemData( THEDP->menusMapper, p_obj, i_val_type,
             val, psz_var );
+
+    /* remove previous signal-slot connection(s) if any */
+    action->disconnect( );
+
     CONNECT( action, triggered(), THEDP->menusMapper, map() );
     THEDP->menusMapper->setMapping( action, itemData );
 
-- 
1.5.6.5

