From d30fc3d6d588d51ab9db67eef753f8d16881f0b4 Mon Sep 17 00:00:00 2001
From: Nicolas Chauvet <kwizart@gmail.com>
Date: Fri, 21 Sep 2018 15:45:21 +0200
Subject: [PATCH] Workaround a bug with ./vlc-cache-gen on armhfp

---
 bin/Makefile.am | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/bin/Makefile.am b/bin/Makefile.am
index 2877e8b0ea..521111e734 100644
--- a/bin/Makefile.am
+++ b/bin/Makefile.am
@@ -142,7 +142,7 @@ endif
 ../modules/plugins.dat: vlc-cache-gen$(EXEEXT)
 	$(AM_V_at)rm -f ../modules/plugins.dat
 	$(AM_V_GEN)if test "$(build)" = "$(host)"; then \
-		./vlc-cache-gen$(EXEEXT) `realpath ../modules` ; \
+		./vlc-cache-gen$(EXEEXT) `realpath ../modules` || : ; \
 	else \
 		echo "Cross-compilation: cache generation skipped!" ; \
 	fi
-- 
2.17.1

