From e2a0e38258a16da56a5d850e179a691839558dcd Mon Sep 17 00:00:00 2001
From: Dominik 'Rathann' Mierzejewski <dominik@greysector.net>
Date: Sun, 4 Sep 2011 23:58:53 +0200
Subject: [PATCH] port to FFmpeg-0.8 API

Signed-off-by: Nicolas Chauvet <kwizart@gmail.com>
---
 modules/stream_out/switcher.c |   41 ++++++++++++++++++++++++++++++++++++++++-
 1 files changed, 40 insertions(+), 1 deletions(-)

diff --git a/modules/stream_out/switcher.c b/modules/stream_out/switcher.c
index ca57a28..d96f690 100644
--- a/modules/stream_out/switcher.c
+++ b/modules/stream_out/switcher.c
@@ -358,6 +358,7 @@ static sout_stream_id_t *Add( sout_stream_t *p_stream, es_format_t *p_fmt )
         /* Set CPU capabilities */
         unsigned i_cpu = vlc_CPU();
         id->ff_enc_c->dsp_mask = 0;
+#if FF_API_MM_FLAGS
         if( !(i_cpu & CPU_CAPABILITY_MMX) )
         {
             id->ff_enc_c->dsp_mask |= FF_MM_MMX;
@@ -375,7 +376,25 @@ static sout_stream_id_t *Add( sout_stream_t *p_stream, es_format_t *p_fmt )
             id->ff_enc_c->dsp_mask |= FF_MM_SSE;
             id->ff_enc_c->dsp_mask |= FF_MM_SSE2;
         }
-
+#else
+        if( !(i_cpu & CPU_CAPABILITY_MMX) )
+        {
+            id->ff_enc_c->dsp_mask |= AV_CPU_FLAG_MMX;
+        }
+        if( !(i_cpu & CPU_CAPABILITY_MMXEXT) )
+        {
+            id->ff_enc_c->dsp_mask |= AV_CPU_FLAG_MMX2;
+        }
+        if( !(i_cpu & CPU_CAPABILITY_3DNOW) )
+        {
+            id->ff_enc_c->dsp_mask |= AV_CPU_FLAG_3DNOW;
+        }
@@ -764,6 +784,25 @@ static mtime_t VideoCommand( sout_stream_t *p_stream, sout_stream_id_t *id )
             id->ff_enc_c->dsp_mask |= FF_MM_SSE;
             id->ff_enc_c->dsp_mask |= FF_MM_SSE2;
         }
+#else
+        if( !(i_cpu & CPU_CAPABILITY_MMX) )
+        {
+            id->ff_enc_c->dsp_mask |= AV_CPU_FLAG_MMX;
+        }
+        if( !(i_cpu & CPU_CAPABILITY_MMXEXT) )
+        {
+            id->ff_enc_c->dsp_mask |= AV_CPU_FLAG_MMX2;
+        }
+        if( !(i_cpu & CPU_CAPABILITY_3DNOW) )
+        {
+            id->ff_enc_c->dsp_mask |= AV_CPU_FLAG_3DNOW;
+        }
+        if( !(i_cpu & CPU_CAPABILITY_SSE) )
+        {
+            id->ff_enc_c->dsp_mask |= AV_CPU_FLAG_SSE;
+            id->ff_enc_c->dsp_mask |= AV_CPU_FLAG_SSE2;
+        }
+#endif

         id->ff_enc_c->width = p_sys->p_pictures[p_sys->i_cmd-1].format.i_width;
         id->ff_enc_c->height = p_sys->p_pictures[p_sys->i_cmd-1].format.i_height;
-- 
1.7.4.4

