From d3bef7cde676ad96ce036cb10aa534ca1c5d0e24 Mon Sep 17 00:00:00 2001
From: Nicolas Chauvet (kwizart) <kwizart@gmail.com>
Date: Tue, 26 May 2009 18:38:07 +0200
Subject: [PATCH] Default libv4l2 to true

---
 modules/access/v4l2.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/modules/access/v4l2.c b/modules/access/v4l2.c
index 31dae66..42a8e9d 100644
--- a/modules/access/v4l2.c
+++ b/modules/access/v4l2.c
@@ -261,7 +261,7 @@ vlc_module_begin ()
     add_integer( CFG_PREFIX "caching", DEFAULT_PTS_DELAY / 1000, NULL,
                 CACHING_TEXT, CACHING_LONGTEXT, true )
 #ifdef HAVE_LIBV4L2
-    add_bool( CFG_PREFIX "use-libv4l2", false, NULL, LIBV4L2_TEXT, LIBV4L2_LONGTEXT, true );
+    add_bool( CFG_PREFIX "use-libv4l2", true, NULL, LIBV4L2_TEXT, LIBV4L2_LONGTEXT, true );
 #endif
 
     set_section( N_( "Tuner" ), NULL )
-- 
1.6.0.6

