--- a/modules/codec/dav1d.c
+++ b/modules/codec/dav1d.c
@@ -223,7 +223,7 @@
         if( p_data )
         {
             res = dav1d_send_data(p_sys->c, p_data);
-            if (res < 0 && res != DAV1D_ERR(EAGAIN))
+            if (res < 0 && res != -EAGAIN)
             {
                 msg_Err(dec, "Decoder feed error %d!", res);
                 i_ret = VLC_EGENERIC;
@@ -248,15 +248,14 @@
             decoder_QueueVideo(dec, pic);
             dav1d_picture_unref(&img);
         }
-        else if (res != DAV1D_ERR(EAGAIN))
+        else if (res != -EAGAIN)
         {
             msg_Warn(dec, "Decoder error %d!", res);
             break;
         }
 
         /* on drain, we must ignore the 1st EAGAIN */
-        if(!b_draining && (res == DAV1D_ERR(EAGAIN) || res == 0)
-                       && (p_data == NULL||b_eos))
+        if(!b_draining && (res == -EAGAIN || res == 0) && (p_data == NULL||b_eos))
         {
             b_draining = true;
             res = 0;

