From fefc0d51b9d3ec6046a73cf317b31870048d1fc2 Mon Sep 17 00:00:00 2001
From: Nicolas Chauvet <kwizart@gmail.com>
Date: Mon, 6 Apr 2020 09:08:08 +0200
Subject: [PATCH] Lower libgcrypt to 1.5.3

Signed-off-by: Nicolas Chauvet <kwizart@gmail.com>
---
 configure.ac | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/configure.ac b/configure.ac
index 1dfe2fce9b..cdac533e10 100644
--- a/configure.ac
+++ b/configure.ac
@@ -4038,14 +4038,14 @@ AC_ARG_ENABLE(libgcrypt,
 AC_ARG_VAR([GCRYPT_CFLAGS], [C compiler flags for gcrypt])
 AC_ARG_VAR([GCRYPT_LIBS], [linker flags flags for gcrypt])
 
-# require libgcrypt >= 1.6.0
+# require libgcrypt >= 1.5.3
 AS_IF([test "${enable_libgcrypt}" != "no"], [
-  AC_CACHE_CHECK([for gcrypt 1.6.0 or later], [ac_cv_lib_gcrypt], [
+  AC_CACHE_CHECK([for gcrypt 1.5.3 or later], [ac_cv_lib_gcrypt], [
     VLC_SAVE_FLAGS
     CFLAGS="${CFLAGS} ${GCRYPT_CFLAGS}"
     AC_COMPILE_IFELSE([AC_LANG_PROGRAM([
 [#include <gcrypt.h>
-#if GCRYPT_VERSION_NUMBER < 0x010600
+#if GCRYPT_VERSION_NUMBER < 0x010503
 #error
 #endif]])], [
       ac_cv_lib_gcrypt=yes
@@ -4059,7 +4059,7 @@ AS_IF([test "${enable_libgcrypt}" != "no"], [
     AC_DEFINE([HAVE_GCRYPT], 1, [Defined if having gcrypt])
   ], [
     AS_IF([test "${enable_libgcrypt}" = "yes"], [
-      AC_MSG_ERROR([libgcrypt version 1.6.0 or higher not found. Install libgcrypt or pass --disable-libgcrypt.])
+      AC_MSG_ERROR([libgcrypt version 1.5.3 or higher not found. Install libgcrypt or pass --disable-libgcrypt.])
     ])
   ])
 ])
-- 
2.25.1

