diff --git a/modules/access/zip/zipstream.c b/modules/access/zip/zipstream.c
index 0250312..59b3f74 100644
--- a/modules/access/zip/zipstream.c
+++ b/modules/access/zip/zipstream.c
@@ -205,8 +205,8 @@ int StreamOpen( vlc_object_t *p_this )
     if( !p_sys->zipFile )
     {
         msg_Warn( s, "unable to open file" );
-        free( p_sys );
         free( p_sys->fileFunctions );
+        free( p_sys );
         return VLC_EGENERIC;
     }
 
@@ -214,8 +214,8 @@ int StreamOpen( vlc_object_t *p_this )
     char *psz_tmp;
     if( asprintf( &psz_tmp, "%s.xspf", s->psz_path ) == -1 )
     {
-        free( p_sys );
         free( p_sys->fileFunctions );
+        free( p_sys );
         return VLC_ENOMEM;
     }
     p_sys->psz_path = s->psz_path;
@@ -310,7 +310,7 @@ static int Control( stream_t *s, int i_query, va_list args )
                 return VLC_EGENERIC;
             else
             {
-                p_sys->i_len = (size_t) i_position;
+                p_sys->i_pos = (size_t) i_position;
                 return VLC_SUCCESS;
             }
         }

diff --git a/modules/gui/qt4/main_interface.cpp b/modules/gui/qt4/main_interface.cpp
index 97aca1a..31b990a 100644
--- a/modules/gui/qt4/main_interface.cpp
+++ b/modules/gui/qt4/main_interface.cpp
@@ -1134,8 +1134,8 @@ void MainInterface::dropEventPlay( QDropEvent *event, bool b_play )
 
         if( s.length() > 0 ) {
             playlist_Add( THEPL, qtu(s), NULL,
-                          PLAYLIST_APPEND | (first ? PLAYLIST_GO: 0),
-                          PLAYLIST_END, true, false );
+                          PLAYLIST_APPEND | (first ? PLAYLIST_GO: PLAYLIST_PREPARSE),
+                          PLAYLIST_END, true, pl_Unlocked );
             first = false;
             RecentsMRL::getInstance( p_intf )->addRecent( s );
         }

