diff -up vlc-0.8.6f/configure.ac.all_plugin vlc-0.8.6f/configure.ac
--- vlc-0.8.6f/configure.ac.all_plugin	2008-05-07 12:00:49.000000000 +0200
+++ vlc-0.8.6f/configure.ac	2008-05-07 12:00:49.000000000 +0200
@@ -1365,7 +1365,7 @@ dnl
 if test "${target_cpu}" = "i686" -o "${target_cpu}" = "i586" -o "${target_cpu}" = "x86" -o "${target_cpu}" = "i386" -o "${target_cpu}" = "x86_64"
 then
     ARCH="${ARCH} mmx"
-    VLC_ADD_BUILTINS([${ACCEL_MODULES}])
+    VLC_ADD_PLUGINS([${ACCEL_MODULES}])
 fi
 
 dnl
@@ -1390,9 +1390,9 @@ dnl
 AC_ARG_ENABLE(altivec,
 [  --disable-altivec       disable AltiVec optimizations (default enabled on PPC)],
 [ if test "${enable_altivec}" = "yes"; then ARCH="${ARCH} altivec";
-    VLC_ADD_BUILTINS([${ACCEL_MODULES}]) fi ],
+    VLC_ADD_PLUGINS([${ACCEL_MODULES}]) fi ],
 [ if test "${target_cpu}" = "powerpc"; then ARCH="${ARCH} altivec";
-    VLC_ADD_BUILTINS([${ACCEL_MODULES}]) fi ])
+    VLC_ADD_PLUGINS([${ACCEL_MODULES}]) fi ])
 
 dnl
 dnl  Debugging mode
@@ -1904,7 +1904,7 @@ then
       AC_CHECK_HEADERS(dvbpsi/dr.h,
         [ VLC_ADD_PLUGINS([ts])
           if test "${enable_sout}" != "no"; then
-            VLC_ADD_BUILTINS([mux_ts])
+            VLC_ADD_PLUGINS([mux_ts])
           fi
           VLC_ADD_LDFLAGS([mux_ts ts dvb],[-ldvbpsi]) ],
         [  AC_MSG_WARN([cannot find libdvbpsi headers]) ],
@@ -1930,9 +1930,9 @@ then
       then
         dnl  Use a custom libdvbpsi
         AC_MSG_RESULT(${real_dvbpsi_tree}/src/.libs/libdvbpsi.a)
-        VLC_ADD_BUILTINS([ts])
+        VLC_ADD_PLUGINS([ts])
         if test "${enable_sout}" != "no"; then
-          VLC_ADD_BUILTINS([mux_ts])
+          VLC_ADD_PLUGINS([mux_ts])
         fi
         VLC_ADD_CPPFLAGS([mux_ts ts dvb],[-I${real_dvbpsi_tree}/src])
         VLC_ADD_LDFLAGS([mux_ts ts dvb],[${real_dvbpsi_tree}/src/.libs/libdvbpsi.a])
@@ -2483,7 +2483,7 @@ then
       VLC_ADD_LDFLAGS([mpgatofixed32],[-L${real_mad_tree}/.libs])
       LDFLAGS="${LDFLAGS_save} ${LDFLAGS_mpgatofixed32}"
       AC_CHECK_LIB(mad, mad_bit_init, [
-        VLC_ADD_BUILTINS([mpgatofixed32])
+        VLC_ADD_PLUGINS([mpgatofixed32])
         VLC_ADD_LDFLAGS([mpgatofixed32],[-lmad])
         ],[ AC_MSG_ERROR([the specified tree hasn't been compiled ])
       ],[])
@@ -5416,7 +5404,7 @@ then
     eval "${plugin}_p=yes"
   done
 else
-  VLC_ADD_BUILTINS([${PLUGINS}])
+  VLC_ADD_PLUGINS([${PLUGINS}])
   PLUGINS=""
 fi
 AM_CONDITIONAL(HAVE_PLUGINS, ${plugin_support})
diff -up vlc-0.8.6f/vlc-config.in.in.all_plugin vlc-0.8.6f/vlc-config.in.in
--- vlc-0.8.6f/vlc-config.in.in.all_plugin	2008-05-07 12:22:25.000000000 +0200
+++ vlc-0.8.6f/vlc-config.in.in	2008-05-07 12:23:00.000000000 +0200
@@ -209,6 +209,7 @@ while test $# -gt 0; do
     activex)
       ;;
     pic)
+      cppflags="${cppflags} -fPIC -DPIC"
       ;;
     external)
       echo_external=yes
diff -up vlc-0.8.6f/modules/video_chroma/Makefile.am.fpicplugin vlc-0.8.6f/modules/video_chroma/Makefile.am
--- vlc-0.8.6f/modules/video_chroma/Makefile.am.all_plugin	2008-05-07 12:41:42.000000000 +0200
+++ vlc-0.8.6f/modules/video_chroma/Makefile.am	2008-05-07 12:44:42.000000000 +0200
@@ -92,13 +92,13 @@ endif
 
 libi420_rgb_plugin_a_SOURCES = $(SOURCES_i420_rgb)
 libi420_rgb_plugin_la_SOURCES = $(SOURCES_i420_rgb)
-libi420_rgb_plugin_a_CFLAGS = `$(VLC_CONFIG) --cflags plugin i420_rgb`
-libi420_rgb_plugin_a_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin i420_rgb`
-libi420_rgb_plugin_a_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin i420_rgb`
-libi420_rgb_plugin_la_CFLAGS = `$(VLC_CONFIG) --cflags plugin i420_rgb`
-libi420_rgb_plugin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin i420_rgb`
-libi420_rgb_plugin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin i420_rgb`
-libi420_rgb_plugin_la_LDFLAGS = `$(VLC_CONFIG) --libs plugin i420_rgb` \
+libi420_rgb_plugin_a_CFLAGS = `$(VLC_CONFIG) --cflags plugin pic i420_rgb`
+libi420_rgb_plugin_a_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin pic i420_rgb`
+libi420_rgb_plugin_a_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin pic i420_rgb`
+libi420_rgb_plugin_la_CFLAGS = `$(VLC_CONFIG) --cflags plugin pic i420_rgb`
+libi420_rgb_plugin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin pic i420_rgb`
+libi420_rgb_plugin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin pic i420_rgb`
+libi420_rgb_plugin_la_LDFLAGS = `$(VLC_CONFIG) --libs plugin pic i420_rgb` \
 	-rpath '$(libvlcdir)' -avoid-version -module -shrext $(LIBEXT)
 libi420_rgb_plugin_la_LIBADD = $(LTLIBVLC)
 
@@ -107,21 +107,21 @@ libi420_rgb_builtin_la_SOURCES = $(SOURC
 libi420_rgb_a_CFLAGS = `$(VLC_CONFIG) --cflags builtin pic i420_rgb`
 libi420_rgb_a_CXXFLAGS = `$(VLC_CONFIG) --cxxflags builtin pic i420_rgb`
 libi420_rgb_a_OBJCFLAGS = `$(VLC_CONFIG) --objcflags builtin pic i420_rgb`
-libi420_rgb_builtin_la_CFLAGS = `$(VLC_CONFIG) --cflags builtin i420_rgb`
-libi420_rgb_builtin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags builtin i420_rgb`
-libi420_rgb_builtin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags builtin i420_rgb`
+libi420_rgb_builtin_la_CFLAGS = `$(VLC_CONFIG) --cflags builtin pic i420_rgb`
+libi420_rgb_builtin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags builtin pic i420_rgb`
+libi420_rgb_builtin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags builtin pic i420_rgb`
 
 # The i420_rgb_mmx plugin
 
 libi420_rgb_mmx_plugin_a_SOURCES = $(SOURCES_i420_rgb_mmx)
 libi420_rgb_mmx_plugin_la_SOURCES = $(SOURCES_i420_rgb_mmx)
-libi420_rgb_mmx_plugin_a_CFLAGS = `$(VLC_CONFIG) --cflags plugin i420_rgb_mmx`
-libi420_rgb_mmx_plugin_a_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin i420_rgb_mmx`
-libi420_rgb_mmx_plugin_a_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin i420_rgb_mmx`
-libi420_rgb_mmx_plugin_la_CFLAGS = `$(VLC_CONFIG) --cflags plugin i420_rgb_mmx`
-libi420_rgb_mmx_plugin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin i420_rgb_mmx`
-libi420_rgb_mmx_plugin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin i420_rgb_mmx`
-libi420_rgb_mmx_plugin_la_LDFLAGS = `$(VLC_CONFIG) --libs plugin i420_rgb_mmx` \
+libi420_rgb_mmx_plugin_a_CFLAGS = `$(VLC_CONFIG) --cflags plugin pic i420_rgb_mmx`
+libi420_rgb_mmx_plugin_a_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin pic i420_rgb_mmx`
+libi420_rgb_mmx_plugin_a_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin pic i420_rgb_mmx`
+libi420_rgb_mmx_plugin_la_CFLAGS = `$(VLC_CONFIG) --cflags plugin pic i420_rgb_mmx`
+libi420_rgb_mmx_plugin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin pic i420_rgb_mmx`
+libi420_rgb_mmx_plugin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin pic i420_rgb_mmx`
+libi420_rgb_mmx_plugin_la_LDFLAGS = `$(VLC_CONFIG) --libs plugin pic i420_rgb_mmx` \
 	-rpath '$(libvlcdir)' -avoid-version -module -shrext $(LIBEXT)
 libi420_rgb_mmx_plugin_la_LIBADD = $(LTLIBVLC)
 
@@ -130,21 +130,21 @@ libi420_rgb_mmx_builtin_la_SOURCES = $(S
 libi420_rgb_mmx_a_CFLAGS = `$(VLC_CONFIG) --cflags builtin pic i420_rgb_mmx`
 libi420_rgb_mmx_a_CXXFLAGS = `$(VLC_CONFIG) --cxxflags builtin pic i420_rgb_mmx`
 libi420_rgb_mmx_a_OBJCFLAGS = `$(VLC_CONFIG) --objcflags builtin pic i420_rgb_mmx`
-libi420_rgb_mmx_builtin_la_CFLAGS = `$(VLC_CONFIG) --cflags builtin i420_rgb_mmx`
-libi420_rgb_mmx_builtin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags builtin i420_rgb_mmx`
-libi420_rgb_mmx_builtin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags builtin i420_rgb_mmx`
+libi420_rgb_mmx_builtin_la_CFLAGS = `$(VLC_CONFIG) --cflags builtin pic i420_rgb_mmx`
+libi420_rgb_mmx_builtin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags builtin pic i420_rgb_mmx`
+libi420_rgb_mmx_builtin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags builtin pic i420_rgb_mmx`
 
 # The i420_yuy2 plugin
 
 libi420_yuy2_plugin_a_SOURCES = $(SOURCES_i420_yuy2)
 libi420_yuy2_plugin_la_SOURCES = $(SOURCES_i420_yuy2)
-libi420_yuy2_plugin_a_CFLAGS = `$(VLC_CONFIG) --cflags plugin i420_yuy2`
-libi420_yuy2_plugin_a_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin i420_yuy2`
-libi420_yuy2_plugin_a_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin i420_yuy2`
-libi420_yuy2_plugin_la_CFLAGS = `$(VLC_CONFIG) --cflags plugin i420_yuy2`
-libi420_yuy2_plugin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin i420_yuy2`
-libi420_yuy2_plugin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin i420_yuy2`
-libi420_yuy2_plugin_la_LDFLAGS = `$(VLC_CONFIG) --libs plugin i420_yuy2` \
+libi420_yuy2_plugin_a_CFLAGS = `$(VLC_CONFIG) --cflags plugin pic i420_yuy2`
+libi420_yuy2_plugin_a_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin pic i420_yuy2`
+libi420_yuy2_plugin_a_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin pic i420_yuy2`
+libi420_yuy2_plugin_la_CFLAGS = `$(VLC_CONFIG) --cflags plugin pic i420_yuy2`
+libi420_yuy2_plugin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin pic i420_yuy2`
+libi420_yuy2_plugin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin pic i420_yuy2`
+libi420_yuy2_plugin_la_LDFLAGS = `$(VLC_CONFIG) --libs plugin pic i420_yuy2` \
 	-rpath '$(libvlcdir)' -avoid-version -module -shrext $(LIBEXT)
 libi420_yuy2_plugin_la_LIBADD = $(LTLIBVLC)
 
@@ -153,21 +153,21 @@ libi420_yuy2_builtin_la_SOURCES = $(SOUR
 libi420_yuy2_a_CFLAGS = `$(VLC_CONFIG) --cflags builtin pic i420_yuy2`
 libi420_yuy2_a_CXXFLAGS = `$(VLC_CONFIG) --cxxflags builtin pic i420_yuy2`
 libi420_yuy2_a_OBJCFLAGS = `$(VLC_CONFIG) --objcflags builtin pic i420_yuy2`
-libi420_yuy2_builtin_la_CFLAGS = `$(VLC_CONFIG) --cflags builtin i420_yuy2`
-libi420_yuy2_builtin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags builtin i420_yuy2`
-libi420_yuy2_builtin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags builtin i420_yuy2`
+libi420_yuy2_builtin_la_CFLAGS = `$(VLC_CONFIG) --cflags builtin pic i420_yuy2`
+libi420_yuy2_builtin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags builtin pic i420_yuy2`
+libi420_yuy2_builtin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags builtin pic i420_yuy2`
 
 # The i420_yuy2_mmx plugin
 
 libi420_yuy2_mmx_plugin_a_SOURCES = $(SOURCES_i420_yuy2_mmx)
 libi420_yuy2_mmx_plugin_la_SOURCES = $(SOURCES_i420_yuy2_mmx)
-libi420_yuy2_mmx_plugin_a_CFLAGS = `$(VLC_CONFIG) --cflags plugin i420_yuy2_mmx`
-libi420_yuy2_mmx_plugin_a_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin i420_yuy2_mmx`
-libi420_yuy2_mmx_plugin_a_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin i420_yuy2_mmx`
-libi420_yuy2_mmx_plugin_la_CFLAGS = `$(VLC_CONFIG) --cflags plugin i420_yuy2_mmx`
-libi420_yuy2_mmx_plugin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin i420_yuy2_mmx`
-libi420_yuy2_mmx_plugin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin i420_yuy2_mmx`
-libi420_yuy2_mmx_plugin_la_LDFLAGS = `$(VLC_CONFIG) --libs plugin i420_yuy2_mmx` \
+libi420_yuy2_mmx_plugin_a_CFLAGS = `$(VLC_CONFIG) --cflags plugin pic i420_yuy2_mmx`
+libi420_yuy2_mmx_plugin_a_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin pic i420_yuy2_mmx`
+libi420_yuy2_mmx_plugin_a_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin pic i420_yuy2_mmx`
+libi420_yuy2_mmx_plugin_la_CFLAGS = `$(VLC_CONFIG) --cflags plugin pic i420_yuy2_mmx`
+libi420_yuy2_mmx_plugin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin pic i420_yuy2_mmx`
+libi420_yuy2_mmx_plugin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin pic i420_yuy2_mmx`
+libi420_yuy2_mmx_plugin_la_LDFLAGS = `$(VLC_CONFIG) --libs plugin pic i420_yuy2_mmx` \
 	-rpath '$(libvlcdir)' -avoid-version -module -shrext $(LIBEXT)
 libi420_yuy2_mmx_plugin_la_LIBADD = $(LTLIBVLC)
 
@@ -176,21 +176,21 @@ libi420_yuy2_mmx_builtin_la_SOURCES = $(
 libi420_yuy2_mmx_a_CFLAGS = `$(VLC_CONFIG) --cflags builtin pic i420_yuy2_mmx`
 libi420_yuy2_mmx_a_CXXFLAGS = `$(VLC_CONFIG) --cxxflags builtin pic i420_yuy2_mmx`
 libi420_yuy2_mmx_a_OBJCFLAGS = `$(VLC_CONFIG) --objcflags builtin pic i420_yuy2_mmx`
-libi420_yuy2_mmx_builtin_la_CFLAGS = `$(VLC_CONFIG) --cflags builtin i420_yuy2_mmx`
-libi420_yuy2_mmx_builtin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags builtin i420_yuy2_mmx`
-libi420_yuy2_mmx_builtin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags builtin i420_yuy2_mmx`
+libi420_yuy2_mmx_builtin_la_CFLAGS = `$(VLC_CONFIG) --cflags builtin pic i420_yuy2_mmx`
+libi420_yuy2_mmx_builtin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags builtin pic i420_yuy2_mmx`
+libi420_yuy2_mmx_builtin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags builtin pic i420_yuy2_mmx`
 
 # The i420_yuy2_altivec plugin
 
 libi420_yuy2_altivec_plugin_a_SOURCES = $(SOURCES_i420_yuy2_altivec)
 libi420_yuy2_altivec_plugin_la_SOURCES = $(SOURCES_i420_yuy2_altivec)
-libi420_yuy2_altivec_plugin_a_CFLAGS = `$(VLC_CONFIG) --cflags plugin i420_yuy2_altivec`
-libi420_yuy2_altivec_plugin_a_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin i420_yuy2_altivec`
-libi420_yuy2_altivec_plugin_a_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin i420_yuy2_altivec`
-libi420_yuy2_altivec_plugin_la_CFLAGS = `$(VLC_CONFIG) --cflags plugin i420_yuy2_altivec`
-libi420_yuy2_altivec_plugin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin i420_yuy2_altivec`
-libi420_yuy2_altivec_plugin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin i420_yuy2_altivec`
-libi420_yuy2_altivec_plugin_la_LDFLAGS = `$(VLC_CONFIG) --libs plugin i420_yuy2_altivec` \
+libi420_yuy2_altivec_plugin_a_CFLAGS = `$(VLC_CONFIG) --cflags plugin pic i420_yuy2_altivec`
+libi420_yuy2_altivec_plugin_a_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin pic i420_yuy2_altivec`
+libi420_yuy2_altivec_plugin_a_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin pic i420_yuy2_altivec`
+libi420_yuy2_altivec_plugin_la_CFLAGS = `$(VLC_CONFIG) --cflags plugin pic i420_yuy2_altivec`
+libi420_yuy2_altivec_plugin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin pic i420_yuy2_altivec`
+libi420_yuy2_altivec_plugin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin pic i420_yuy2_altivec`
+libi420_yuy2_altivec_plugin_la_LDFLAGS = `$(VLC_CONFIG) --libs plugin pic i420_yuy2_altivec` \
 	-rpath '$(libvlcdir)' -avoid-version -module -shrext $(LIBEXT)
 libi420_yuy2_altivec_plugin_la_LIBADD = $(LTLIBVLC)
 
@@ -199,21 +199,21 @@ libi420_yuy2_altivec_builtin_la_SOURCES 
 libi420_yuy2_altivec_a_CFLAGS = `$(VLC_CONFIG) --cflags builtin pic i420_yuy2_altivec`
 libi420_yuy2_altivec_a_CXXFLAGS = `$(VLC_CONFIG) --cxxflags builtin pic i420_yuy2_altivec`
 libi420_yuy2_altivec_a_OBJCFLAGS = `$(VLC_CONFIG) --objcflags builtin pic i420_yuy2_altivec`
-libi420_yuy2_altivec_builtin_la_CFLAGS = `$(VLC_CONFIG) --cflags builtin i420_yuy2_altivec`
-libi420_yuy2_altivec_builtin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags builtin i420_yuy2_altivec`
-libi420_yuy2_altivec_builtin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags builtin i420_yuy2_altivec`
+libi420_yuy2_altivec_builtin_la_CFLAGS = `$(VLC_CONFIG) --cflags builtin pic i420_yuy2_altivec`
+libi420_yuy2_altivec_builtin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags builtin pic i420_yuy2_altivec`
+libi420_yuy2_altivec_builtin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags builtin pic i420_yuy2_altivec`
 
 # The i422_yuy2 plugin
 
 libi422_yuy2_plugin_a_SOURCES = $(SOURCES_i422_yuy2)
 libi422_yuy2_plugin_la_SOURCES = $(SOURCES_i422_yuy2)
-libi422_yuy2_plugin_a_CFLAGS = `$(VLC_CONFIG) --cflags plugin i422_yuy2`
-libi422_yuy2_plugin_a_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin i422_yuy2`
-libi422_yuy2_plugin_a_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin i422_yuy2`
-libi422_yuy2_plugin_la_CFLAGS = `$(VLC_CONFIG) --cflags plugin i422_yuy2`
-libi422_yuy2_plugin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin i422_yuy2`
-libi422_yuy2_plugin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin i422_yuy2`
-libi422_yuy2_plugin_la_LDFLAGS = `$(VLC_CONFIG) --libs plugin i422_yuy2` \
+libi422_yuy2_plugin_a_CFLAGS = `$(VLC_CONFIG) --cflags plugin pic i422_yuy2`
+libi422_yuy2_plugin_a_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin pic i422_yuy2`
+libi422_yuy2_plugin_a_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin pic i422_yuy2`
+libi422_yuy2_plugin_la_CFLAGS = `$(VLC_CONFIG) --cflags plugin pic i422_yuy2`
+libi422_yuy2_plugin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin pic i422_yuy2`
+libi422_yuy2_plugin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin pic i422_yuy2`
+libi422_yuy2_plugin_la_LDFLAGS = `$(VLC_CONFIG) --libs plugin pic i422_yuy2` \
 	-rpath '$(libvlcdir)' -avoid-version -module -shrext $(LIBEXT)
 libi422_yuy2_plugin_la_LIBADD = $(LTLIBVLC)
 
@@ -222,21 +222,21 @@ libi422_yuy2_builtin_la_SOURCES = $(SOUR
 libi422_yuy2_a_CFLAGS = `$(VLC_CONFIG) --cflags builtin pic i422_yuy2`
 libi422_yuy2_a_CXXFLAGS = `$(VLC_CONFIG) --cxxflags builtin pic i422_yuy2`
 libi422_yuy2_a_OBJCFLAGS = `$(VLC_CONFIG) --objcflags builtin pic i422_yuy2`
-libi422_yuy2_builtin_la_CFLAGS = `$(VLC_CONFIG) --cflags builtin i422_yuy2`
-libi422_yuy2_builtin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags builtin i422_yuy2`
-libi422_yuy2_builtin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags builtin i422_yuy2`
+libi422_yuy2_builtin_la_CFLAGS = `$(VLC_CONFIG) --cflags builtin pic i422_yuy2`
+libi422_yuy2_builtin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags builtin pic i422_yuy2`
+libi422_yuy2_builtin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags builtin pic i422_yuy2`
 
 # The i422_yuy2_mmx plugin
 
 libi422_yuy2_mmx_plugin_a_SOURCES = $(SOURCES_i422_yuy2_mmx)
 libi422_yuy2_mmx_plugin_la_SOURCES = $(SOURCES_i422_yuy2_mmx)
-libi422_yuy2_mmx_plugin_a_CFLAGS = `$(VLC_CONFIG) --cflags plugin i422_yuy2_mmx`
-libi422_yuy2_mmx_plugin_a_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin i422_yuy2_mmx`
-libi422_yuy2_mmx_plugin_a_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin i422_yuy2_mmx`
-libi422_yuy2_mmx_plugin_la_CFLAGS = `$(VLC_CONFIG) --cflags plugin i422_yuy2_mmx`
-libi422_yuy2_mmx_plugin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin i422_yuy2_mmx`
-libi422_yuy2_mmx_plugin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin i422_yuy2_mmx`
-libi422_yuy2_mmx_plugin_la_LDFLAGS = `$(VLC_CONFIG) --libs plugin i422_yuy2_mmx` \
+libi422_yuy2_mmx_plugin_a_CFLAGS = `$(VLC_CONFIG) --cflags plugin pic i422_yuy2_mmx`
+libi422_yuy2_mmx_plugin_a_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin pic i422_yuy2_mmx`
+libi422_yuy2_mmx_plugin_a_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin pic i422_yuy2_mmx`
+libi422_yuy2_mmx_plugin_la_CFLAGS = `$(VLC_CONFIG) --cflags plugin pic i422_yuy2_mmx`
+libi422_yuy2_mmx_plugin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin pic i422_yuy2_mmx`
+libi422_yuy2_mmx_plugin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin pic i422_yuy2_mmx`
+libi422_yuy2_mmx_plugin_la_LDFLAGS = `$(VLC_CONFIG) --libs plugin pic i422_yuy2_mmx` \
 	-rpath '$(libvlcdir)' -avoid-version -module -shrext $(LIBEXT)
 libi422_yuy2_mmx_plugin_la_LIBADD = $(LTLIBVLC)
 
@@ -245,21 +245,21 @@ libi422_yuy2_mmx_builtin_la_SOURCES = $(
 libi422_yuy2_mmx_a_CFLAGS = `$(VLC_CONFIG) --cflags builtin pic i422_yuy2_mmx`
 libi422_yuy2_mmx_a_CXXFLAGS = `$(VLC_CONFIG) --cxxflags builtin pic i422_yuy2_mmx`
 libi422_yuy2_mmx_a_OBJCFLAGS = `$(VLC_CONFIG) --objcflags builtin pic i422_yuy2_mmx`
-libi422_yuy2_mmx_builtin_la_CFLAGS = `$(VLC_CONFIG) --cflags builtin i422_yuy2_mmx`
-libi422_yuy2_mmx_builtin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags builtin i422_yuy2_mmx`
-libi422_yuy2_mmx_builtin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags builtin i422_yuy2_mmx`
+libi422_yuy2_mmx_builtin_la_CFLAGS = `$(VLC_CONFIG) --cflags builtin pic i422_yuy2_mmx`
+libi422_yuy2_mmx_builtin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags builtin pic i422_yuy2_mmx`
+libi422_yuy2_mmx_builtin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags builtin pic i422_yuy2_mmx`
 
 # The i420_ymga plugin
 
 libi420_ymga_plugin_a_SOURCES = $(SOURCES_i420_ymga)
 libi420_ymga_plugin_la_SOURCES = $(SOURCES_i420_ymga)
-libi420_ymga_plugin_a_CFLAGS = `$(VLC_CONFIG) --cflags plugin i420_ymga`
-libi420_ymga_plugin_a_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin i420_ymga`
-libi420_ymga_plugin_a_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin i420_ymga`
-libi420_ymga_plugin_la_CFLAGS = `$(VLC_CONFIG) --cflags plugin i420_ymga`
-libi420_ymga_plugin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin i420_ymga`
-libi420_ymga_plugin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin i420_ymga`
-libi420_ymga_plugin_la_LDFLAGS = `$(VLC_CONFIG) --libs plugin i420_ymga` \
+libi420_ymga_plugin_a_CFLAGS = `$(VLC_CONFIG) --cflags plugin pic i420_ymga`
+libi420_ymga_plugin_a_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin pic i420_ymga`
+libi420_ymga_plugin_a_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin pic i420_ymga`
+libi420_ymga_plugin_la_CFLAGS = `$(VLC_CONFIG) --cflags plugin pic i420_ymga`
+libi420_ymga_plugin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin pic i420_ymga`
+libi420_ymga_plugin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin pic i420_ymga`
+libi420_ymga_plugin_la_LDFLAGS = `$(VLC_CONFIG) --libs plugin pic i420_ymga` \
 	-rpath '$(libvlcdir)' -avoid-version -module -shrext $(LIBEXT)
 libi420_ymga_plugin_la_LIBADD = $(LTLIBVLC)
 
@@ -268,21 +268,21 @@ libi420_ymga_builtin_la_SOURCES = $(SOUR
 libi420_ymga_a_CFLAGS = `$(VLC_CONFIG) --cflags builtin pic i420_ymga`
 libi420_ymga_a_CXXFLAGS = `$(VLC_CONFIG) --cxxflags builtin pic i420_ymga`
 libi420_ymga_a_OBJCFLAGS = `$(VLC_CONFIG) --objcflags builtin pic i420_ymga`
-libi420_ymga_builtin_la_CFLAGS = `$(VLC_CONFIG) --cflags builtin i420_ymga`
-libi420_ymga_builtin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags builtin i420_ymga`
-libi420_ymga_builtin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags builtin i420_ymga`
+libi420_ymga_builtin_la_CFLAGS = `$(VLC_CONFIG) --cflags builtin pic i420_ymga`
+libi420_ymga_builtin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags builtin pic i420_ymga`
+libi420_ymga_builtin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags builtin pic i420_ymga`
 
 # The i420_ymga_mmx plugin
 
 libi420_ymga_mmx_plugin_a_SOURCES = $(SOURCES_i420_ymga_mmx)
 libi420_ymga_mmx_plugin_la_SOURCES = $(SOURCES_i420_ymga_mmx)
-libi420_ymga_mmx_plugin_a_CFLAGS = `$(VLC_CONFIG) --cflags plugin i420_ymga_mmx`
-libi420_ymga_mmx_plugin_a_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin i420_ymga_mmx`
-libi420_ymga_mmx_plugin_a_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin i420_ymga_mmx`
-libi420_ymga_mmx_plugin_la_CFLAGS = `$(VLC_CONFIG) --cflags plugin i420_ymga_mmx`
-libi420_ymga_mmx_plugin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin i420_ymga_mmx`
-libi420_ymga_mmx_plugin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin i420_ymga_mmx`
-libi420_ymga_mmx_plugin_la_LDFLAGS = `$(VLC_CONFIG) --libs plugin i420_ymga_mmx` \
+libi420_ymga_mmx_plugin_a_CFLAGS = `$(VLC_CONFIG) --cflags plugin pic i420_ymga_mmx`
+libi420_ymga_mmx_plugin_a_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin pic i420_ymga_mmx`
+libi420_ymga_mmx_plugin_a_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin pic i420_ymga_mmx`
+libi420_ymga_mmx_plugin_la_CFLAGS = `$(VLC_CONFIG) --cflags plugin pic i420_ymga_mmx`
+libi420_ymga_mmx_plugin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags plugin pic i420_ymga_mmx`
+libi420_ymga_mmx_plugin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags plugin pic i420_ymga_mmx`
+libi420_ymga_mmx_plugin_la_LDFLAGS = `$(VLC_CONFIG) --libs plugin pic i420_ymga_mmx` \
 	-rpath '$(libvlcdir)' -avoid-version -module -shrext $(LIBEXT)
 libi420_ymga_mmx_plugin_la_LIBADD = $(LTLIBVLC)
 
@@ -291,7 +291,7 @@ libi420_ymga_mmx_builtin_la_SOURCES = $(
 libi420_ymga_mmx_a_CFLAGS = `$(VLC_CONFIG) --cflags builtin pic i420_ymga_mmx`
 libi420_ymga_mmx_a_CXXFLAGS = `$(VLC_CONFIG) --cxxflags builtin pic i420_ymga_mmx`
 libi420_ymga_mmx_a_OBJCFLAGS = `$(VLC_CONFIG) --objcflags builtin pic i420_ymga_mmx`
-libi420_ymga_mmx_builtin_la_CFLAGS = `$(VLC_CONFIG) --cflags builtin i420_ymga_mmx`
-libi420_ymga_mmx_builtin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags builtin i420_ymga_mmx`
-libi420_ymga_mmx_builtin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags builtin i420_ymga_mmx`
+libi420_ymga_mmx_builtin_la_CFLAGS = `$(VLC_CONFIG) --cflags builtin pic i420_ymga_mmx`
+libi420_ymga_mmx_builtin_la_CXXFLAGS = `$(VLC_CONFIG) --cxxflags builtin pic i420_ymga_mmx`
+libi420_ymga_mmx_builtin_la_OBJCFLAGS = `$(VLC_CONFIG) --objcflags builtin pic i420_ymga_mmx`
 
