diff -up vlc-26fb05e95724bd19f9a144c51bc11f79c609f3bc/configure.ac.vpl vlc-26fb05e95724bd19f9a144c51bc11f79c609f3bc/configure.ac
--- vlc-26fb05e95724bd19f9a144c51bc11f79c609f3bc/configure.ac.vpl	2023-06-17 17:19:28.954540887 +0200
+++ vlc-26fb05e95724bd19f9a144c51bc11f79c609f3bc/configure.ac	2023-06-17 17:19:28.956540899 +0200
@@ -2925,7 +2925,7 @@ fi
 dnl
 dnl Intel QuickSync (aka MediaSDK) H264/H262 encoder
 dnl
-PKG_ENABLE_MODULES_VLC([MFX], [qsv], [libmfx], [Intel QuickSync MPEG4-Part10/MPEG2 (aka H.264/H.262) encoder], [auto])
+PKG_ENABLE_MODULES_VLC([MFX], [qsv], [vpl], [Intel QuickSync MPEG4-Part10/MPEG2 (aka H.264/H.262) encoder], [auto])
 
 dnl
 dnl libfluidsynth (MIDI synthetizer) plugin
diff -up vlc-26fb05e95724bd19f9a144c51bc11f79c609f3bc/modules/codec/qsv.c.vpl vlc-26fb05e95724bd19f9a144c51bc11f79c609f3bc/modules/codec/qsv.c
--- vlc-26fb05e95724bd19f9a144c51bc11f79c609f3bc/modules/codec/qsv.c.vpl	2023-06-09 17:29:29.000000000 +0200
+++ vlc-26fb05e95724bd19f9a144c51bc11f79c609f3bc/modules/codec/qsv.c	2023-06-17 17:22:05.468382540 +0200
@@ -33,7 +33,7 @@
 #include <vlc_picture.h>
 #include <vlc_codec.h>
 
-#include <mfx/mfxvideo.h>
+#include <vpl/mfxvideo.h>
 
 #define SOUT_CFG_PREFIX     "sout-qsv-"
 
