diff -up vlc-0.8.6f/configure.ac.all_plugin vlc-0.8.6f/configure.ac
--- vlc-0.8.6f/configure.ac.wx28iccfixes	2008-05-07 11:28:24.000000000 +0200
+++ vlc-0.8.6f/configure.ac	2008-05-07 11:30:08.000000000 +0200
@@ -4204,7 +4187,7 @@ then
     AC_TRY_COMPILE([#define ALSA_PCM_NEW_HW_PARAMS_API
                     #define ALSA_PCM_NEW_SW_PARAMS_API
                     #include <alsa/asoundlib.h>],
-       [void foo() { snd_pcm_hw_params_get_period_time(0,0,0); }],
+       [void foo(); { snd_pcm_hw_params_get_period_time(0,0,0); }],
         AC_DEFINE(HAVE_ALSA_NEW_API, 1, Define if ALSA is at least rc4))
     VLC_ADD_PLUGINS([alsa])
     VLC_ADD_LDFLAGS([alsa],[-lasound -lm -ldl])
diff -up vlc-0.8.6f/include/vlc_common.h.wx28iccfixes vlc-0.8.6f/include/vlc_common.h
--- vlc-0.8.6f/include/vlc_common.h.wx28iccfixes	2008-05-07 11:28:24.000000000 +0200
+++ vlc-0.8.6f/include/vlc_common.h	2008-05-07 11:30:08.000000000 +0200
@@ -604,6 +604,30 @@ static inline int64_t GCD( int64_t a, in
     }                                                                         \
     while( 0 )
 
+/* Special version for C++ code */
+#define REMOVE_ELEM_PP( type_p_ar, p_ar, i_oldsize, i_pos )                   \
+    do                                                                        \
+    {                                                                         \
+        if( (i_oldsize) - (i_pos) - 1 )                                       \
+        {                                                                     \
+            memmove( (p_ar) + (i_pos),                                        \
+                     (p_ar) + (i_pos) + 1,                                    \
+                     ((i_oldsize) - (i_pos) - 1) * sizeof( *(p_ar) ) );       \
+        }                                                                     \
+        if( i_oldsize > 1 )                                                   \
+        {                                                                     \
+            (p_ar) = (type_p_ar)realloc( p_ar, ((i_oldsize) - 1) * sizeof( *(p_ar) ) );  \
+        }                                                                     \
+        else                                                                  \
+        {                                                                     \
+            free( p_ar );                                                     \
+            (p_ar) = NULL;                                                    \
+        }                                                                     \
+        (i_oldsize)--;                                                        \
+    }                                                                         \
+    while( 0 )
+
+
 #define TAB_INIT( count, tab )                  \
   do {                                          \
     (count) = 0;                                \
diff -up vlc-0.8.6f/modules/gui/wxwidgets/dialogs/interaction.cpp.wx28iccfixes vlc-0.8.6f/modules/gui/wxwidgets/dialogs/interaction.cpp
--- vlc-0.8.6f/modules/gui/wxwidgets/dialogs/interaction.cpp.wx28iccfixes	2008-05-07 11:13:14.000000000 +0200
+++ vlc-0.8.6f/modules/gui/wxwidgets/dialogs/interaction.cpp	2008-05-07 11:27:53.000000000 +0200
@@ -245,7 +245,7 @@ void InteractionDialog::OnClear( wxComma
         user_widget_t *p_widget = p_dialog->pp_widgets[i];
         FREE( p_widget->psz_text );
         FREE( p_widget->val.psz_string );
-        REMOVE_ELEM( p_dialog->pp_widgets, p_dialog->i_widgets, i );
+        REMOVE_ELEM_PP( user_widget_t **, p_dialog->pp_widgets, p_dialog->i_widgets, i );
         free( p_widget );
     }
     widgets_panel->DestroyChildren();
diff -up vlc-0.8.6f/modules/gui/wxwidgets/menus.cpp.wx28iccfixes vlc-0.8.6f/modules/gui/wxwidgets/menus.cpp
--- vlc-0.8.6f/modules/gui/wxwidgets/menus.cpp.wx28iccfixes	2008-05-06 22:34:27.000000000 +0200
+++ vlc-0.8.6f/modules/gui/wxwidgets/menus.cpp	2008-05-07 11:02:19.000000000 +0200
@@ -28,7 +28,7 @@
 #include "interface.hpp"
 
 #include <wx/dynarray.h>
-WX_DEFINE_ARRAY(int, ArrayOfInts);
+WX_DEFINE_ARRAY_INT(int, ArrayOfInts);
 WX_DEFINE_ARRAY_PTR(const char *, ArrayOfStrings);
 
 
