diff --git a/modules/access/http.c b/modules/access/http.c
index 66e32ff..c465fa4 100644
--- a/modules/access/http.c
+++ b/modules/access/http.c
@@ -452,7 +452,13 @@ connect:
         if( i_ret == DIALOG_OK_YES )
         {
             msg_Dbg( p_access, "retrying with user=%s, pwd=%s",
-                        psz_login, psz_password );
+                     psz_login,
+#if 1
+                     "yeah right, like we're going to print a password."
+#else
+                     psz_password
+#endif
+                        );
             if( psz_login ) p_sys->url.psz_username = strdup( psz_login );
             if( psz_password ) p_sys->url.psz_password = strdup( psz_password );
             free( psz_login );
diff --git a/modules/codec/avcodec/video.c b/modules/codec/avcodec/video.c
index 3fa837e..0e90b9f 100644
--- a/modules/codec/avcodec/video.c
+++ b/modules/codec/avcodec/video.c
@@ -301,6 +301,8 @@ int InitVideoDec( decoder_t *p_dec, AVCodecContext *p_context,
         p_sys->p_context->pix_fmt != PIX_FMT_YUV422P &&
         /* H264 uses too many reference frames */
         p_sys->i_codec_id != CODEC_ID_H264 &&
+        /* No idea why ... but this fixes flickering on some TSCC streams */
+        p_sys->i_codec_id != CODEC_ID_TSCC &&
         !p_sys->p_context->debug_mv )
     {
         /* Some codecs set pix_fmt only after the 1st frame has been decoded,
diff --git a/modules/codec/x264.c b/modules/codec/x264.c
index 68508c9..04fce89 100644
--- a/modules/codec/x264.c
+++ b/modules/codec/x264.c
@@ -75,7 +75,7 @@ static void Close( vlc_object_t * );
     "I-frames are inserted only every other keyint frames, which probably " \
     "leads to ugly encoding artifacts. Range 1 to 100." )
 
-#if X264_BUILD >= 55 /* r607 */
+#if X264_BUILD >= 55 /* r607 */ && X264_BUILD < 67 /* r1117 */
 #define PRESCENE_TEXT N_("Faster, less precise scenecut detection" )
 #define PRESCENE_LONGTEXT N_( "Faster, less precise scenecut detection. " \
     "Required and implied by multi-threading." )
@@ -429,8 +429,12 @@ vlc_module_begin();
         change_integer_range( -1, 100 );
 
 #if X264_BUILD >= 55 /* r607 */
+#  if X264_BUILD < 67 /* r1117 */
     add_bool( SOUT_CFG_PREFIX "pre-scenecut", 0, NULL, PRESCENE_TEXT,
               PRESCENE_LONGTEXT, false );
+#  else
+    add_obsolete_bool( "pre-scenecut" )
+#  endif
 #endif
 
     add_integer( SOUT_CFG_PREFIX "bframes", 0, NULL, BFRAMES_TEXT,
@@ -957,9 +961,12 @@ static int  Open ( vlc_object_t *p_this )
         p_sys->param.i_scenecut_threshold = val.i_int;
 #endif
 
-#if X264_BUILD >= 55 /* r607 */
+#if X264_BUILD >= 55 /* r607 */ && X264_BUILD < 67 /* r1117 */
     var_Get( p_enc, SOUT_CFG_PREFIX "pre-scenecut", &val );
     p_sys->param.b_pre_scenecut = val.b_bool;
+#endif
+
+#if X264_BUILD >= 55 /* r607 */
     var_Get( p_enc, SOUT_CFG_PREFIX "non-deterministic", &val );
     p_sys->param.b_deterministic = val.b_bool;
 #endif
