diff -up vlc-0.8.6h/configure.ac.compat vlc-0.8.6h/configure.ac
--- vlc-0.8.6h/configure.ac.compat	2008-06-07 17:28:53.000000000 +0200
+++ vlc-0.8.6h/configure.ac	2008-06-07 17:30:16.000000000 +0200
@@ -2703,15 +2714,15 @@ dnl Look for a ffmpeg-config (we are on 
  else
 
 dnl Trying with pkg-config
-   PKG_CHECK_MODULES(FFMPEG,[libavcodec, libavformat],
+   PKG_CHECK_MODULES(FFMPEG,[libavcodec-compat, libavformat-compat],
     [
      VLC_SAVE_FLAGS
      CPPFLAGS="${CPPFLAGS} ${FFMPEG_CFLAGS}"
      CFLAGS="${CFLAGS} ${FFMPEG_CFLAGS}"
-     AC_CHECK_HEADERS(ffmpeg/avcodec.h libavcodec/avcodec.h)
-     AC_CHECK_HEADERS(ffmpeg/avformat.h libavformat/avformat.h)
-     AC_CHECK_HEADERS(ffmpeg/avutil.h libavutil/avutil.h)
-     AC_CHECK_HEADERS(postproc/postprocess.h libpostproc/postprocess.h)
+     AC_CHECK_HEADERS(avcodec.h ffmpeg/avcodec.h)
+     AC_CHECK_HEADERS(avformat.h ffmpeg/avformat.h)
+     AC_CHECK_HEADERS(avutil.h ffmpeg/avutil.h)
+     AC_CHECK_HEADERS(postprocess.h postproc/postprocess.h)
      if test "${SYS}" = "darwin"; then
         VLC_ADD_BUILTINS([ffmpeg])
      else
@@ -2727,7 +2738,7 @@ dnl Trying with pkg-config
      VLC_ADD_CFLAGS([ffmpeg stream_out_switcher],[${FFMPEG_CFLAGS}])
      VLC_ADD_LDFLAGS([ffmpeg stream_out_switcher],[${FFMPEG_LIBS}])
      dnl newer ffmpeg have a separate libpostproc
-     PKG_CHECK_MODULES(POSTPROC, libpostproc,[
+     PKG_CHECK_MODULES(POSTPROC, libpostproc-compat,[
        VLC_ADD_LDFLAGS([ffmpeg],[${POSTPROC_LIBS}]) 
        ],[ true ])
      VLC_RESTORE_FLAGS
@@ -2776,8 +2787,8 @@ then
      VLC_SAVE_FLAGS
      CFLAGS="${CFLAGS} ${CFLAGS_ffmpeg}"
      LDFLAGS="${LDFLAGS} ${LDFLAGS_ffmpeg}"
-     AC_CHECK_LIB(avcodec, img_resample, ,
-       [AC_MSG_ERROR([Your FFmpeg library doesn't have the needed img_resample() function. You should rebuild it with software scaler disabled])], -lavutil)
+     AC_CHECK_LIB(avcodec-compat, img_resample, ,
+       [AC_MSG_ERROR([Your FFmpeg library doesn't have the needed img_resample() function. You should rebuild it with software scaler disabled])], -lavutil-compat)
      VLC_RESTORE_FLAGS
 fi
 
