diff -up vlc-1.1.6/modules/visualization/projectm.cpp.hardode_path vlc-1.1.6/modules/visualization/projectm.cpp
--- vlc-1.1.6/modules/visualization/projectm.cpp.hardode_path   2011-01-17 17:30:54.000000000 +0100
+++ vlc-1.1.6/modules/visualization/projectm.cpp        2011-01-25 10:55:30.791607002 +0100
@@ -89,9 +89,9 @@ vlc_module_begin ()
     add_directory( "projectm-preset-path", "/usr/share/projectM/presets", NULL,
 #endif
                    PRESET_PATH_TXT, PRESET_PATH_LONGTXT, true )
-    add_file( "projectm-title-font", "/usr/share/fonts/truetype/ttf-dejavu/DejaVuSans.ttf", NULL,
+    add_file( "projectm-title-font", "/usr/share/fonts/dejavu/DejaVuSans.ttf", NULL,
               TITLE_FONT_TXT, TITLE_FONT_LONGTXT, true )
-    add_file( "projectm-menu-font", "/usr/share/fonts/truetype/ttf-dejavu/DejaVuSansMono.ttf", NULL,
+    add_file( "projectm-menu-font", "/usr/share/fonts/dejavu/DejaVuSansMono.ttf", NULL,
               MENU_FONT_TXT, MENU_FONT_LONGTXT, true )
 #endif
     add_integer( "projectm-width", 800, NULL, WIDTH_TEXT, WIDTH_LONGTEXT,
