diff -up vlc-0.8.6c/modules/stream_out/switcher.c.fix_for_compat vlc-0.8.6c/modules/stream_out/switcher.c
--- vlc-0.8.6c/modules/stream_out/switcher.c.fix_for_compat	2007-09-21 02:21:35.000000000 +0200
+++ vlc-0.8.6c/modules/stream_out/switcher.c	2007-09-21 02:21:45.000000000 +0200
@@ -45,7 +45,7 @@
 #ifdef HAVE_POSTPROC_POSTPROCESS_H
 #   include <postproc/postprocess.h>
 #else
-#   include <libpostproc/postprocess.h>
+#   include <postprocess.h>
 #endif
 
 #define SOUT_CFG_PREFIX "sout-switcher-"
diff -up vlc-0.8.6c/modules/codec/ffmpeg/ffmpeg.c.fix_for_compat vlc-0.8.6c/modules/codec/ffmpeg/ffmpeg.c
--- vlc-0.8.6c/modules/codec/ffmpeg/ffmpeg.c.fix_for_compat	2007-06-16 16:25:12.000000000 +0200
+++ vlc-0.8.6c/modules/codec/ffmpeg/ffmpeg.c	2007-09-21 02:15:18.000000000 +0200
@@ -46,7 +46,7 @@
 #   ifdef HAVE_POSTPROC_POSTPROCESS_H
 #       include <postproc/postprocess.h>
 #   else
-#       include <libpostproc/postprocess.h>
+#       include <postprocess.h>
 #   endif
 #endif
 
diff -up vlc-0.8.6c/modules/codec/ffmpeg/postprocess.c.fix_for_compat vlc-0.8.6c/modules/codec/ffmpeg/postprocess.c
--- vlc-0.8.6c/modules/codec/ffmpeg/postprocess.c.fix_for_compat	2007-06-16 16:25:12.000000000 +0200
+++ vlc-0.8.6c/modules/codec/ffmpeg/postprocess.c	2007-09-21 02:15:18.000000000 +0200
@@ -40,7 +40,7 @@
 #ifdef HAVE_POSTPROC_POSTPROCESS_H
 #   include <postproc/postprocess.h>
 #else
-#   include <libpostproc/postprocess.h>
+#   include <postprocess.h>
 #endif
 
 #ifndef PP_CPU_CAPS_ALTIVEC
