diff -up vlc-0.8.6f/mozilla/vlcshell.cpp.libxul vlc-0.8.6f/mozilla/vlcshell.cpp
--- vlc-0.8.6f/mozilla/vlcshell.cpp.libxul	2008-03-23 23:41:49.000000000 +0100
+++ vlc-0.8.6f/mozilla/vlcshell.cpp	2008-04-01 02:50:58.000000000 +0200
@@ -36,9 +36,7 @@
 #endif
 
 /* This is from mozilla java, do we really need it? */
-#if 0
 #include <jri.h>
-#endif
 
 #include "vlcplugin.h"
 
@@ -325,7 +323,7 @@ NPError NPP_Destroy( NPP instance, NPSav
 
     instance->pdata = NULL;
 
-#if XP_WIN
+#ifdef XP_WIN
     HWND win = (HWND)p_plugin->getWindow()->window;
     WNDPROC winproc = p_plugin->getWindowProc();
     if( winproc )
@@ -652,7 +650,7 @@ void NPP_Print( NPP instance, NPPrint* p
 /******************************************************************************
  * Windows-only methods
  *****************************************************************************/
-#if XP_WIN
+#ifdef XP_WIN
 static LRESULT CALLBACK Manage( HWND p_hwnd, UINT i_msg, WPARAM wpar, LPARAM lpar )
 {
     VlcPlugin* p_plugin = reinterpret_cast<VlcPlugin*>(GetWindowLongPtr(p_hwnd, GWLP_USERDATA));
diff -up vlc-0.8.6f/mozilla/vlcplugin.cpp.libxul vlc-0.8.6f/mozilla/vlcplugin.cpp
--- vlc-0.8.6f/mozilla/vlcplugin.cpp.libxul	2008-03-23 23:41:49.000000000 +0100
+++ vlc-0.8.6f/mozilla/vlcplugin.cpp	2008-04-01 02:48:41.000000000 +0200
@@ -50,10 +50,10 @@ VlcPlugin::VlcPlugin( NPP instance, uint
     p_scriptClass(NULL),
     p_browser(instance),
     psz_baseURL(NULL)
-#if XP_WIN
+#ifdef XP_WIN
     ,pf_wndproc(NULL)
 #endif
-#if XP_UNIX
+#ifdef XP_UNIX
     ,i_width((unsigned)-1)
     ,i_height((unsigned)-1)
 #endif
@@ -448,7 +448,7 @@ relativeurl:
     return NULL;
 }
 
-#if XP_UNIX
+#ifdef XP_UNIX
 int  VlcPlugin::setSize(unsigned width, unsigned height)
 {
     int diff = (width != i_width) || (height != i_height);
diff -up vlc-0.8.6f/mozilla/vlcplugin.h.libxul vlc-0.8.6f/mozilla/vlcplugin.h
--- vlc-0.8.6f/mozilla/vlcplugin.h.libxul	2008-03-23 23:41:49.000000000 +0100
+++ vlc-0.8.6f/mozilla/vlcplugin.h	2008-04-01 04:00:33.000000000 +0200
@@ -78,14 +78,14 @@ public:
                             { libvlc_log = log; };
     libvlc_log_t*       getLog()
                             { return libvlc_log; };
-#if XP_WIN
+#ifdef XP_WIN
     WNDPROC             getWindowProc()
                             { return pf_wndproc; };
     void                setWindowProc(WNDPROC wndproc)
                             { pf_wndproc = wndproc; };
 #endif
 
-#if XP_UNIX
+#ifdef XP_UNIX
     int                 setSize(unsigned width, unsigned height);
 #endif
 
@@ -108,10 +108,10 @@ private:
 
     /* display settings */
     NPWindow  npwindow;
-#if XP_WIN
+#ifdef XP_WIN
     WNDPROC   pf_wndproc;
 #endif
-#if XP_UNIX
+#ifdef XP_UNIX
     unsigned int     i_width, i_height;
 #endif
 };
diff -up vlc-0.8.6f/configure.ac.libxul vlc-0.8.6f/configure.ac
--- vlc-0.8.6f/configure.ac.libxul	2008-04-01 02:48:40.000000000 +0200
+++ vlc-0.8.6f/configure.ac	2008-04-01 02:48:41.000000000 +0200
@@ -5119,7 +5119,7 @@ then
   if test "${with_mozilla_sdk_path}" = "" -o "${with_mozilla_sdk_path}" = "no"
   then
     dnl New firefox-1.5 SDK api
-    PKG_CHECK_MODULES([FIREFOX], [firefox-plugin],
+    PKG_CHECK_MODULES([FIREFOX], [libxul],
       [
         PKG_CHECK_MODULES([FIREFOX_XPCOM], [firefox-xpcom],
           [],
@@ -5138,7 +5138,6 @@ then
           ])
         CPPFLAGS="${CPPFLAGS_save} ${FIREFOX_CFLAGS} ${FIREFOX_XPCOM_CFLAGS} ${MOZILLA_NSPR_CFLAGS} ${NSPR_CFLAGS} ${MOZILLA_CPPFLAGS}"
         MOZILLA_REQUIRED_HEADERS=1
-        AC_CHECK_HEADERS(mozilla-config.h,,MOZILLA_REQUIRED_HEADERS=0)
         AC_CHECK_HEADERS(npapi.h,,MOZILLA_REQUIRED_HEADERS=0)
         AC_CHECK_HEADERS(npruntime.h,,MOZILLA_REQUIRED_HEADERS=0,
           [#if HAVE_NPAPI_H
@@ -5150,21 +5160,14 @@ then
            #endif
           ])
         if test "${MOZILLA_REQUIRED_HEADERS}" = "0"; then
-            AC_MSG_ERROR([Please install the Firefox development tools; mozilla-config.h, plugin/npapi.h and plugin/npruntime.h were not found.])
+            AC_MSG_ERROR([Please install the Firefox development tools; plugin/npapi.h and plugin/npruntime.h were not found.])
         fi
+        AC_DEFINE([XP_UNIX],[1],[Defined for Mozilla as we are Unix])
+        AC_DEFINE([MOZ_X11],[1],[Defined as Unix mozilla uses X11])
+        AC_DEFINE([OJI],[1],[Java Runtime Interface])
         MOZILLA_REQUIRED_HEADERS=
         mozilla=:
-        firefox_include=`pkg-config --variable=includedir firefox-plugin`
-        if test -f "${firefox_include}/mozilla-config.h" -a -f "${firefox_include}/mozilla-config32.h"; then
-            MOZILLA_CONFIG_H="${firefox_include}/mozilla-config32.h"
-        elif test -f "${firefox_include}/mozilla-config.h" -a -f "${firefox_include}/mozilla-config64.h"; then
-            MOZILLA_CONFIG_H="${firefox_include}/mozilla-config64.h"
-        else
-            MOZILLA_CONFIG_H="${firefox_include}/mozilla-config.h"
-        fi
-        if grep '^#define MOZ_X11 1' ${MOZILLA_CONFIG_H} 2>&1 > /dev/null ; then
-            VLC_ADD_LDFLAGS([mozilla], [${X_LIBS} ${X_PRE_LIBS} -lX11 -lXt])
-        fi
+        VLC_ADD_LDFLAGS([mozilla], [${X_LIBS} ${X_PRE_LIBS} -lX11 -lXt])
         VLC_ADD_CPPFLAGS([mozilla],[${CPPFLAGS}])
         VLC_ADD_LDFLAGS([mozilla],[${FIREFOX_LIBS} ${MOZILLA_NSPR_LIBS} ${NSPR_LIBS} ${MOZILLA_LDFLAGS}])
         VLC_ADD_PLUGINS([mozilla])
