diff -up vlc-2.0.4/bin/Makefile.am.cache vlc-2.0.4/bin/Makefile.am
--- vlc-2.0.4/bin/Makefile.am.cache	2011-12-08 19:00:25.000000000 +0100
+++ vlc-2.0.4/bin/Makefile.am	2012-10-19 23:14:23.302736639 +0200
@@ -80,7 +80,5 @@ MOSTLYCLEANFILES = $(noinst_DATA)
 ../modules/plugins.dat: vlc-cache-gen$(EXEEXT)
 	$(AM_V_at)rm -f ../modules/plugins.dat
 	$(AM_V_GEN)if test "$(build)" = "$(host)"; then \
-		./vlc-cache-gen$(EXEEXT) ../modules ; \
-	else \
-		echo "Cross-compilation: cache generation skipped!" ; \
+		echo "Builder compilation: cache generation skipped!" ; \
 	fi
