commit 942ab6c31026a84e4c5503cb8e68a5135dfc2f2b
Author: Pierre Ynard <linkfanel@yahoo.fr>
Date:   Sun Jan 23 18:19:05 2011 +0100

    lirc: build fix

diff --git a/modules/control/lirc.c b/modules/control/lirc.c
index aee33c3..e8c4780 100644
--- a/modules/control/lirc.c
+++ b/modules/control/lirc.c
@@ -25,6 +25,7 @@
  * Preamble
  *****************************************************************************/
 
+#include <errno.h>
 #include <fcntl.h>
 
 #ifdef HAVE_CONFIG_H
@@ -154,10 +155,12 @@
         /* Wait for data */
         struct pollfd ufd = { .fd = p_sys->i_fd, .events = POLLIN, .revents = 0 };
         if( poll( &ufd, 1, -1 ) == -1 )
+        {
             if( errno == EINTR )
                 continue;
             else
                 break;
+        }
 
         /* Process */
         int canc = vlc_savecancel();
