commit de68b12937589cdcbe5143f162dabd7654969362
Author: Rémi Denis-Courmont <remi@remlab.net>
Date:   Sat Jul 24 13:43:56 2010 +0300

    Do not override signal functions for the time being

diff --git a/bin/override.c b/bin/override.c
index 2385f2b..ac9e14c 100644
--- a/bin/override.c
+++ b/bin/override.c
@@ -201,6 +201,7 @@ int rand (void)
 }
 
 
+#if 0
 /** Signals **/
 #include <signal.h>
 
@@ -252,6 +253,7 @@ error:
     LOG("Blocked", "%d, %p, %p", signum, act, old);
     return -1;
 }
+#endif
 
 
 /*** Locales ***

